### ISSUE TITLE
Bug: overloads not matched when using `TypeVar` with `default` (where Pyright matches correctly)

### ISSUE THREAD

**Issue by MarcoGorelli** [python/mypy#19182]
**Bug Report**

(A clear and concise description of what the bug is.)

**To Reproduce**

```python
from typing import Any, Callable, Generic, Protocol, Self, overload, reveal_type

from typing_extensions import TypeVar

T = TypeVar("T", bound=str | int, default=Any)


class Foo(Generic[T]):
    def __init__(self, t: T) -> None:
        self.t = t


class Bar(Protocol):
    @overload
    def apply(self, f: Callable[..., Foo]) -> Self: ...
    @overload
    def apply(self, f: Callable[..., T | None]) -> Foo[T]: ...


def func() -> Foo:
    return Foo(3)


def main(b: Bar) -> None:
    reveal_type(b.apply(func))

```

https://mypy-play.net/?mypy=latest&python=3.12&gist=6df33ff32887ddfab4c1ae43edd62b2d

**Expected Behavior**

that `reveal_type` would show `Bar` (for reference, this is what Pyright does)

**Actual Behavior**

```
note: Revealed type is "Any"
```

**Your Environment**

- Mypy version used: 1.15.0
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.12

**Comment by erictraut** [comment]
In the minimal producer without overloads, mypy doesn't detect a type error for the call expression `b.apply(func)`, but I think [it should](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAMICGANtZQEbUCmZA4kykyEgMYBQfUJFiJUaAPpMAHjA4BnJGBRzM2XPgAqiJgDVKIARqgBeKFoS79ACgBEGm2XpgArigAmxuTBBQAPphQYMjcmYEpnahhjIjgASgE%2BHjo5FQAxMDArdk5uHgBtDQBdWIAuPigKqBDgKHFxVCQYOqs5JmpgMhgSs1ioAFoAPigAOSUmMsrJqFb2gDp8UxgEpMoUqAAhfQnK6qhKBARqOBa2jqhgbqpaBmY82fuyI39RzmL%2BofSwAsLu%2B9mE3bAVw8Ky9QZQT7bCogJgwZwgFAQjJWADM8QEuwglFQVno3U2IDBQxe43KlXos32h2OQJQPF6UAAxFAAMoACxc1DcUHoTD2iK44AMQA). That seems to be the root of the problem here.

The argument `func` has the type `Callable[[], Foo[Any]]`. For the call to succeed, this type must be assignable to `Callable[..., T | None]` which means `Foo[Any]` must be assignable to `T | None`. Since `T` has an upper bound of `str | int`, that means `Foo[Any]` must be assignable to `str | int | None`, which it is not. That means there's a type violation. This is independent of whether `T` can be solved or not, so the default value shouldn't come into play here.

If I remove the default from the type parameter `T`, mypy correctly detects this error.

**Comment by sterliakov** [comment]
@erictraut `Any` **is** a valid solution for `T`. `mypy` does not suggest it by default, but using `default=Any` typevars makes `mypy` use `Any` as explicitly allowed value. With `T = Any`, the call is `def apply(f: Callable[..., Any | None]) -> Foo[Any]`, and thus `apply(func)` is allowed.

While I probably agree that using default only when there were no constraints at all would be more reasonable, `[]` represents both "no constraints" and "constraints can't be built" internally, and that won't be an easy fix. I'm not sure whether `default=Any` is worth such work - essentially this is just another escape hatch, and `Any` is rather contagious anyway.

**Comment by sterliakov** [comment]
From the spec:

> type checkers may use a type parameter’s default when the type parameter cannot be solved to anything.

I think mypy is conformant here: there's no solution for `T` in the OP example, so `mypy` is correct to use the provided default of `Any`?

Your last example is convincing, though: it produces a diagnostic even if `T` is just `default=Any`:

```python
T = TypeVar("T", bound=Sequence[int], default=Any)

def func(x: T) -> T: ...

func([1])  # OK
func([""])  # E: Value of type variable "T" of "func" cannot be "list[str]"  [type-var]
```

So our approach is indeed at least inconsistent.

**Comment by MarcoGorelli** [comment]
Thanks all for responses

> Could you elaborate on your use case where default=Any makes sense?

In pandas, `Series` is generic in the data type. For example, `Series[int]`, `Series[float]`, .... There are cases when the inner type can't be determined statically due to value-dependent behaviour (😭), and so `Series[Any]` is used. I was exploring whether defaulting `Series` to `Series[Any]` was feasible

**Comment by jorenham** [comment]
> Could you elaborate on your use case where `default=Any` makes sense? Usually it's a bad idea, `mypy` produces a diagnostic when generic's typevar is not filled, and `default=Any` essentially just disables that check.

In NumPy this is used quite a lot, for example

- `floating` defaults to `floating[Any]`, (since 2.1)
- `dtype` defaults to `dtype[Any]` (since 2.3), and
- `ndarray` defaults to `ndarray[tuple[Any, ...], dtype[Any]]` (since 2.3)

There are also some examples of this in the typeshed stdlib stubs, e.g. `builtins.slice`.
